plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt

```diff
@@ -267,84 +274,92 @@ class CameraAnimationsPluginImpl : CameraAnimationsPlugin {
     animator.addInternalListener(object : Animator.AnimatorListener {
 
       override fun onAnimationStart(animation: Animator) {
-        (animation as? CameraAnimator<*>)?.apply {
-          // check for a specific use-case when canceling an animation with start delay that
-          // has not yet started - in that case onAnimationStart logic must be skipped
-          if (canceled) {
-            return
-          }
-          lifecycleListeners.forEach {
-            it.onAnimatorStarting(type, this, owner)
-          }
-          mapTransformDelegate.setUserAnimationInProgress(true)
-          // check if such animation is not running already
-          // if it is - then cancel it
-          // Safely iterate over new set because of the possible changes of "this.animators" in Animator callbacks
-          HashSet(animators).forEach {
-            if (it.type == type && it.isRunning && it != this) {
-              lifecycleListeners.forEach { listener ->
-                listener.onAnimatorInterrupting(type, it, it.owner, this, this.owner)
+        handler.post {
+          (animation as? CameraAnimator<*>)?.apply {
+            // check for a specific use-case when canceling an animation with start delay that
+            // has not yet started - in that case onAnimationStart logic must be skipped
+            if (canceled) {
+              return@post
+            }
+            lifecycleListeners.forEach {
+              it.onAnimatorStarting(type, this, owner)
+            }
+            mapTransformDelegate.setUserAnimationInProgress(true)
```

Comment:
just to double check: if we create beforehand `val mainHandler = Handler(Looper.getMainLooper())` and will actually call native functions here as `mainHandler.post { mapTransformDelegate.setUserAnimationInProgress(true) }` - will this work on Xiaomi device when device is locked? If it will work - then imo it will be a bit safer to call gl-native functions from main thread while creating animators on background thread.

Reply:
yeah i think it works, but i'm a little unsure about the state lock `setUserAnimationInProgress`. If that is on a different thread than the rest of this function then the `nativeMap.isUserAnimationInProgress` is going to be checking on the wrong thread.

Actually, this makes makes it unnecessary to change between the main thread and the animation thread https://github.com/mapbox/mapbox-maps-android/pull/1605#issuecomment-1217939433

It seems like it would be better to designate an animation thread, and never switch to the main thread (unless you're updating a view)

Reply:
@kmadsen it's not about the thread checking. Core requires native `ALooper` be attached to the thread to allow scheduling in C++ if needed from main thread to other native threads. When you create background handler thread in Android (as done in this PR) - it does have the looper but native code does not make use of it as we're using custom run loop implementation.

As per current architecture attaching native looper is not exposed to Java (so you can't attach it to your background thread) and implemented internally for 2 threads _only_: main Android thread and our Mapbox render thread.

To sum up: e.g. `nativeMap.setCamera` does not make use of any scheduling logic (most likely but who knows 😉) under the hood but potentially some other native function invoked from background handler thread may result in a native crash. IMO it's still safer to actually run animators on background thread in Java but schedule native calls thru `mainHandler` as I described above. In that case ThreadChecker could be actually enabled.

Reply:
> In that case ThreadChecker could be actually enabled.

There are potentially large refactors to ensure all native calls will happen on the main thread. One of them that I found was in the [CameraAnimatorsFactory](https://github.com/mapbox/mapbox-maps-android/blob/2220e35d817837a16c36a976530b57d6d7fd923f/plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimatorsFactory.kt#L367). This part of the animation uses a [TypeEvaluator](https://developer.android.com/reference/android/animation/TypeEvaluator) which is synchronous.

If we were to force main thread there, it will also require a thread context switch in the middle of the animation. There may be a bunch of cases like this that will require careful thought before being able to enable ThreadChecker with background thread animations.

Reply:
https://github.com/mapbox/mapbox-maps-android/pull/1605#issuecomment-1218644974

^ Here's another example from the location animations where it crashes

Reply:
> https://github.com/mapbox/mapbox-maps-android/pull/1605#issuecomment-1218644974
> ^ Here's another example from the location animations where it crashes

to fix this one I assume we should switch back to the main thread in the update listener here
https://github.com/mapbox/mapbox-maps-android/blob/main/plugin-locationcomponent/src/main/java/com/mapbox/maps/plugin/locationcomponent/animators/PuckAnimator.kt#L33

Reply:
Overal I agree with @kiryldz - we call gl-native from the main thread for a reason (and even added the thread checker in the debug builds). It seems risky and fragile to change to calling from other thread without native looper attached, but also considering possible races with the state of gl-native. 

@kmadsen would it be possible to just postMain updates in every place where `ValueAnimator.addUpdateListener` is added? E.g. `PuckAnimator` and `CameraAnimator`?

Reply:
> would it be possible to just postMain updates in every place where ValueAnimator.addUpdateListener is added? E.g. PuckAnimator and CameraAnimator?

This was the reason I introduced the `postMain` function. Because, am trying to switch to the main thread everywhere. But that is not a small change, there are many places where gl-native is called. Some implementations may require a large refactor - see this comment https://github.com/mapbox/mapbox-maps-android/pull/1605#discussion_r948208432

> There are potentially large refactors to ensure all native calls will happen on the main thread. One of them that I found was in the [CameraAnimatorsFactory](https://github.com/mapbox/mapbox-maps-android/blob/2220e35d817837a16c36a976530b57d6d7fd923f/plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimatorsFactory.kt#L367). This part of the animation uses a [TypeEvaluator](https://developer.android.com/reference/android/animation/TypeEvaluator) which is synchronous.